File: src/ios/Sources/UpdaterPlugin.swift
import Foundation

/**
 * The platform half of the over-the-air updater: where bundles live, what the
 * app's version is, and an HTTP client.
 *
 * Everything that decides anything — signature and hash checks, staging,
 * install and rollback — is in Rust, where it is the same code on both
 * platforms and is tested on the host. This side only moves bytes.
 *
 * `downloadFromRust` blocks, deliberately, and it is safe to: the Rust side
 * only calls it from the updater's own worker thread, and URLSession delivers
 * the completion on its own delegate queue, never the thread that is waiting,
 * so the wait cannot deadlock whichever thread it is on.
 *
 * Errors come back as a JSON object carrying `error`.
 */
@objc(UpdaterPlugin)
public class UpdaterPlugin: NSObject {

    /// No cache and no cookies: an update check must see the server's answer
    /// now, not one an intermediate cache kept.
    private static let session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: configuration)
    }()

    private static func json(_ object: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: object),
              let text = String(data: data, encoding: .utf8) else {
            return "{\"error\":\"The updater could not encode its answer.\"}"
        }
        return text
    }

    private static func errorJson(_ message: String) -> String {
        json(["error": message])
    }

    /// Under Application Support and excluded from backup: a bundle is a cache
    /// of something the server can always send again, and restoring one onto
    /// a device with a different app build would only have it thrown away as
    /// the wrong runtime.
    @objc
    public func dataDirectoryFromRust() -> String? {
        guard let support = FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first else { return nil }
        var directory = support
            .appendingPathComponent("g3_native_plugins", isDirectory: true)
            .appendingPathComponent("updater", isDirectory: true)
        do {
            try FileManager.default.createDirectory(
                at: directory,
                withIntermediateDirectories: true
            )
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
            try directory.setResourceValues(values)
        } catch {
            NSLog("[UpdaterPlugin] could not prepare the bundle directory: \(error)")
            return nil
        }
        return directory.path
    }

    @objc
    public func appVersionFromRust() -> String? {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
    }

    /// Fetch a URL to a file and report the status. The body is only moved to
    /// the destination for a 2xx other than 204. URLSession downloads to a
    /// temporary file of its own and deletes it once the completion returns,
    /// so the move has to happen inside the completion.
    @objc
    public func downloadFromRust(_ requestJson: String) -> String? {
        guard let data = requestJson.data(using: .utf8),
              let request = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let urlText = request["url"] as? String,
              let url = URL(string: urlText),
              let destinationPath = request["destination"] as? String else {
            return Self.errorJson("Could not read the download request.")
        }
        let destination = URL(fileURLWithPath: destinationPath)
        let timeout = (request["timeoutMs"] as? Double).map { $0 / 1000 } ?? 30
        var urlRequest = URLRequest(url: url, timeoutInterval: timeout)
        for (name, value) in request["headers"] as? [String: String] ?? [:] {
            urlRequest.setValue(value, forHTTPHeaderField: name)
        }

        let done = DispatchSemaphore(value: 0)
        // A box rather than a captured `var`: the completion handler is
        // @Sendable in current SDKs, and mutating a captured local from one is
        // a compile error. The semaphore orders the write before the read.
        let answer = Answer(Self.errorJson("The download did not finish."))
        let task = Self.session.downloadTask(with: urlRequest) { location, response, error in
            defer { done.signal() }
            if let error = error {
                answer.value = Self.errorJson("Download failed: \(error.localizedDescription)")
                return
            }
            guard let http = response as? HTTPURLResponse else {
                answer.value = Self.errorJson("The server's answer was not HTTP.")
                return
            }
            let status = http.statusCode
            if (200..<300).contains(status), status != 204, let location = location {
                do {
                    let manager = FileManager.default
                    try manager.createDirectory(
                        at: destination.deletingLastPathComponent(),
                        withIntermediateDirectories: true
                    )
                    if manager.fileExists(atPath: destination.path) {
                        try manager.removeItem(at: destination)
                    }
                    try manager.moveItem(at: location, to: destination)
                } catch {
                    answer.value = Self.errorJson("Could not move the download into place: \(error)")
                    return
                }
            }
            answer.value = Self.json(["status": status])
        }
        task.resume()
        done.wait()
        return answer.value
    }
}

private final class Answer {
    var value: String
    init(_ value: String) { self.value = value }
}

File: src/android/updater/src/main/kotlin/dev/dioxus/g3_native_plugins/updater/UpdaterPlugin.kt
package dev.dioxus.g3_native_plugins.updater

import android.app.Activity
import android.content.pm.PackageManager
import android.os.Build
import java.io.File
import java.net.HttpURLConnection
import java.net.URI
import org.json.JSONObject

/**
 * The platform half of the over-the-air updater: where bundles live, what the
 * app's version is, and an HTTP client.
 *
 * Everything that decides anything — signature and hash checks, staging,
 * install and rollback — is in Rust, where it is the same code on both
 * platforms and is tested on the host. This side only moves bytes.
 *
 * [downloadFromRust] blocks, deliberately. The Rust side only calls it from
 * the updater's own worker thread, never from a Dioxus thread or the main
 * thread (where Android would refuse network access outright), and a blocking
 * call keeps the whole pipeline one straight line of code over there.
 *
 * Errors come back as a JSON object carrying `error`.
 */
class UpdaterPlugin(private val activity: Activity) {
    // Named errorJson, not error: kotlin.error() throws, and a member named
    // the same would be a very quiet trap for whoever edits this next.
    private fun errorJson(message: String): String =
        JSONObject().put("error", message).toString()

    /**
     * Under `noBackupFilesDir`: a bundle is a cache of something the server
     * can always send again, and restoring one onto a device with a different
     * app build would only have it thrown away as the wrong runtime.
     */
    fun dataDirectoryFromRust(): String {
        val directory = File(activity.noBackupFilesDir, "g3_native_plugins/updater")
        directory.mkdirs()
        return directory.absolutePath
    }

    fun appVersionFromRust(): String? = try {
        val info = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            activity.packageManager.getPackageInfo(
                activity.packageName,
                PackageManager.PackageInfoFlags.of(0),
            )
        } else {
            @Suppress("DEPRECATION")
            activity.packageManager.getPackageInfo(activity.packageName, 0)
        }
        info.versionName
    } catch (_: Exception) {
        null
    }

    /**
     * Fetch a URL to a file and report the status. The body is only written
     * for a 2xx other than 204, and only ever by renaming a finished temporary
     * file into place, so a half-downloaded file never sits at the
     * destination looking complete.
     */
    fun downloadFromRust(requestJson: String): String {
        val request = try {
            JSONObject(requestJson)
        } catch (failure: Exception) {
            return errorJson("Could not read the download request: ${failure.message}")
        }
        val destination = File(request.getString("destination"))
        val partial = File(destination.path + ".part")
        var connection: HttpURLConnection? = null
        return try {
            connection = URI(request.getString("url")).toURL().openConnection() as HttpURLConnection
            val timeout = request.optInt("timeoutMs", 30_000)
            connection.connectTimeout = timeout
            connection.readTimeout = timeout
            // Redirects are followed within a scheme; HttpURLConnection never
            // follows HTTPS down to HTTP, which is the behavior wanted here.
            connection.instanceFollowRedirects = true
            connection.useCaches = false
            request.optJSONObject("headers")?.let { headers ->
                for (name in headers.keys()) {
                    connection.setRequestProperty(name, headers.getString(name))
                }
            }
            val status = connection.responseCode
            if (status in 200..299 && status != 204) {
                destination.parentFile?.mkdirs()
                connection.inputStream.use { input ->
                    partial.outputStream().use { output -> input.copyTo(output) }
                }
                if (destination.exists()) destination.delete()
                if (!partial.renameTo(destination)) {
                    return errorJson("Could not move the download into place")
                }
            }
            JSONObject().put("status", status).toString()
        } catch (failure: Exception) {
            partial.delete()
            errorJson("Download failed: ${failure.message ?: failure.javaClass.simpleName}")
        } finally {
            connection?.disconnect()
        }
    }
}
